tools/criticalstat.py

```diff
@@ -0,0 +1,335 @@
+    }
+    return 0;
+}
+
+static int in_idle(void)
+{
+     u64 *idlep, idle;
+     int idx = 0;
+
+    // Skip event if we're in idle loop
+    idlep = isidle.lookup(&idx);
+    bpf_probe_read(&idle, sizeof(u64), idlep);
+
+    // Skip if we're in idle loop
+    if (idle)
+        return 1;
+    else
+        return 0;
+}
+
+static void reset_state(void)
+{
+    int idx = 0;
+    struct start_data s = {};
+
+    sts.update(&idx, &s);
+}
+
+TRACEPOINT_PROBE(preemptirq, TYPE_disable)
+{
+    int idx = 0;
+    struct start_data s;
+
+    // Handle the case Ienter, CSenter, CSexit, Iexit
+    // Handle the case Ienter, CSenter, Iexit, CSexit
+    if (in_idle()) {
+        bpf_trace_printk(\"disable: In idle, resetting\\n\");
+        reset_state();
+        return 0;
+    }
+
+    u64 ts = bpf_ktime_get_ns();
+
+    bpf_trace_printk(\"Entered new section, setting time to %lu\\n\",
+        (unsigned long)ts);
+    s.idle_skip = 0;
+    s.addr_offs[start_caller_off] = args->caller_offs;
+    s.addr_offs[start_parent_off] = args->parent_offs;
+    s.ts = ts;
+    s.active = 1;
+    bpf_trace_printk(\"Finished storing\\n\", (unsigned long)ts);
+
+    sts.update(&idx, &s);
+    return 0;
+}
+
+TRACEPOINT_PROBE(preemptirq, TYPE_enable)
+{
+    int idx = 0;
+    u64 start_ts, end_ts, diff;
+    struct start_data *stdp, std;
+
+    // Handle the case CSenter, Ienter, CSexit, Iexit
+    // Handle the case Ienter, CSenter, CSexit, Iexit
+    if (in_idle()) {
+        bpf_trace_printk(\"enable: In idle, resetting\\n\");
+        reset_state();
+        return 0;
+    }
+
+    bpf_trace_printk(\"enable: start lookup\\n\");
+
+    stdp = sts.lookup(&idx);
+    bpf_trace_printk(\"enable: start read\\n\");
+    bpf_probe_read(&std, sizeof(struct start_data), stdp);
+
+    // Handle the case Ienter, Csenter, Iexit, Csexit
+    if (!std.active) {
+        reset_state();
+        return 0;
+    }
+
+    // Handle the case CSenter, Ienter, Iexit, CSexit
+    if (std.idle_skip) {
+        reset_state();
+        return 0;
+    }
+
+    bpf_trace_printk(\"enable: start gettime\\n\");
+    end_ts = bpf_ktime_get_ns();
+    start_ts = std.ts;
+
+    if (start_ts > end_ts) {
+        bpf_trace_printk("ERROR: start < end\\n");
+        reset_state();
+        return 0;
+    }
+
+    diff = end_ts - start_ts;
+    bpf_trace_printk(\"Exited section, diff is %lu\\n\", (unsigned long)diff);
+
+    if (diff < DURATION) {
+        reset_state();
+        return 0;
+    }
+
+    u64 id = bpf_get_current_pid_tgid();
+    struct data_t data = {};
+
+    if (bpf_get_current_comm(&data.comm, sizeof(data.comm)) == 0) {
+        data.addrs[start_caller_off] = std.addr_offs[start_caller_off];
+        data.addrs[start_parent_off] = std.addr_offs[start_parent_off];
+        data.addrs[end_caller_off] = args->caller_offs;
+        data.addrs[end_parent_off] = args->parent_offs;
+
+        data.id = id;
+        data.stack_id = stack_traces.get_stackid(args, BPF_F_REUSE_STACKID);
```

Comment:
Please do not use BPF_F_REUSE_STACKID. just use 0. Reusing stack id may have issues since you have pid_tgid associated with this stack id.

Reply:
Yes, it may give wrong results if there's collision :(.

But with your suggestion, what will happen if there is collision? Are we supposed to discard the sample then and not store it in the perf buffer?

Reply:
If there is a collision, setting flag 0 will give a negative stack id (after casting to int type).
User space will then now that the stack is not valid.

An alternative is to use 4.18 introduced `bpf_get_stack()` helper which will give you a list of ip's. and then you can send to user space directly. But since this is only available at 4.18, I guess you have to leave with old mechanism for a while.